File: r/benchmarker.R
reps <- as.numeric(Sys.getenv("LAMP_REPS"))
output_dir <- Sys.getenv("LAMP_OUTPUT_DIR")
l3_size <- as.numeric(Sys.getenv("LAMP_L3_CACHE_SIZE"))
file_out  <- "default"
file_time <- "default_time"

cs <- matrix(data = rnorm(l3_size), nrow = 1, ncol = l3_size)

cache_scrub <- function(){
  for (j in 1:dim(cs)[2]){
    cs[j] <<- cs[j] + 1e-3
  }
}

init_benchmarker <- function(filename){

  file_out  <<- paste(output_dir, filename, ".txt", sep="")
  file_time <<- paste(output_dir, filename, "_timings.txt", sep="")

  cat("algorithm;R;\n", file=file_out, sep=";")
  cat("", file=file_time, sep=";")

  cat(c("Benchmarker initialized:", file_out, file_time, reps, "\n"), sep=" ")
}

benchmark <- function(name, func, ...){

  times <- matrix(0, nrow=1, ncol=reps)
  result <- ""
  for (i in 1:reps){
      res = func(...)
      times[i] <- res[[1]]
      result <- res[[2]]
  }

  cat(c(name, apply(times, 1, min), "\n"), file=file_out, sep=";", append=TRUE)
  cat(c(name, times, "\n"), file=file_time, sep=";", append=TRUE)

  return(result)
}

File: python/benchmarker.py
import numpy as np
import os
import time
import logging

logger = logging.getLogger('Benchmarker')

l3_size = int(os.environ['LAMP_L3_CACHE_SIZE'])
# cs = np.random.randn(1, l3_size)
cs = np.random.randn(1, 1)


def cache_scrub():
    for j in range(cs.shape[0]):
        cs[j] += 1e-3


class Benchmarker:

    def __init__(self, file_name):

        output_dir = str(os.environ['LAMP_OUTPUT_DIR'])

        self.file_timings = output_dir + file_name + '_timings.txt'
        self.file = output_dir + file_name + '.txt'

        self.reps = int(os.environ['LAMP_REPS'])

        with open(self.file, 'w') as f:
            f.write('algorithm;Python\n')
        with open(self.file_timings, 'w') as f:
            f.write('')

    def benchmark(self, name, func, *args):

        times = np.zeros(self.reps)
        logger.debug('Benchmarker runs started.')
        for i in range(self.reps):
            times[i], res = func(*args)
            logger.debug('Bench executed')
        logger.debug('Benchmarker runs finished.')

        logger.debug('Writing output to file.')
        with open(self.file, 'a') as f:
            f.write('{};'.format(name))
            f.write('{} '.format(np.min(times)))
            f.write('\n')
        with open(self.file_timings, 'a') as f:
            f.write('{};'.format(name))
            for item in times:
                f.write('{};'.format(item))
            f.write('\n')
        logger.debug('Output written to file.')
        return res
